### PR TITLE
support custom file name in `gitea dump` command

### PR DESCRIPTION
**Pull request by glaszig**
## what

with this patch you can submit a custom nam for the backup file created by `gitea dump`.

    gitea dump -f gitea-backup.zip

## why

without this patch, when writing a backup script which stores my gitea dump off-site, i have to do some weird gymnastics to get the file `gitea dump` writes to disk because it generates a random file name with the current unix timestamp. like `dump_file=$(ls -t|head -1)` while being sure to be in an empty directory.

with this patch, i can just choose the file's name myself and move on encrypting and uploading it.

## compatibility

this change should not affect anybody as the new parameter is optional and has a default value which does not change the known behavior of the binary.

### CODE REVIEW

File: cmd/dump.go [hunk 1]
```diff
@@ -85,7 +90,7 @@ func runDump(ctx *cli.Context) error {
 
 	dbDump := path.Join(tmpWorkDir, "gitea-db.sql")
 
-	fileName := fmt.Sprintf("gitea-dump-%d.zip", time.Now().Unix())
+	fileName := fmt.Sprintf(ctx.String("file"), time.Now().Unix())
```
Comment (adelowo) [inline cmd/dump.go]: Kindly update the docs.

Comment (adelowo) [inline cmd/dump.go]: Also if the value of `--file` is provided by the user and doesn't have a `%d` format directive, wouldn't that ruin the value of `fileName`. If I remember correctly, when a string doesn't have a format directive, the resulting output is something like `%!(EXTRA int64=1257894000)`.  So in this case, `fileName` will end up being `userSuppliedFileName%!(EXTRA int64=1257894000)` .

**Review by adelowo — CHANGES_REQUESTED** [review]

File: cmd/dump.go [hunk 2]
```diff
@@ -35,6 +35,11 @@ It can be used for backup and capture Gitea server image to send to maintainer`,
 			Value: "custom/conf/app.ini",
 			Usage: "Custom configuration file path",
 		},
+		cli.StringFlag{
+			Name:  "file, f",
```
Comment (adelowo) [inline cmd/dump.go:39]: Kindly update the docs to reflect the new flag

**Review by adelowo — CHANGES_REQUESTED** [review]

File: cmd/dump.go (hunk 2 above)
Comment (glaszig) [reply cmd/dump.go:39]: done.

File: cmd/dump.go (hunk 1 above)
Comment (glaszig) [reply cmd/dump.go]: you, sir, are completely right. updated the logic.

**Comment by jolheiser** [comment]
In my opinion, the default should remain the same with the timestamp, and the override should just override without giving the `%d` timestamp option.  
  
If a user wants to override the filename, they should provide the exact name they want. If they need a timestamp, they can figure out how to generate it via shell (or whatever terminal they use per their OS).

**Comment by glaszig** [comment]
that‘s how it works with this patch.

**Review by lunny — APPROVED** [review]

**Review by adelowo — APPROVED** [review]

File: cmd/dump.go [hunk 3]
```diff
@@ -164,6 +170,16 @@ func runDump(ctx *cli.Context) error {
 	return nil
 }
 
+func getDumpFileName(ctx *cli.Context) string {
+	fmtString := ctx.String("file")
+
+	if strings.Count(fmtString, "%d") > 0 {
+		fmtString = fmt.Sprintf(fmtString, time.Now().Unix())
```
Comment (jolheiser) [inline cmd/dump.go]: In reference to my other comment, personally I think this is unnecessary.  
  
If a user wants to override the file name, they should be able to provide the exact name they want, without relying on the code transforming a `%d`. This code would only transform a single `%d` as well, and if they had any other formatting symbols before/after, they would also end up with a bad filename.  
  
I will not block this PR, however I don't personally approve of this "magic".
(To clarify, I _do_ like this PR. Overriding the filename would be a great feature. I just don't particularly care for this part.

Comment (jolheiser) [inline cmd/dump.go]: In reference to my other comment, personally I think this is unnecessary.

If a user wants to override the file name, they should be able to provide the exact name they want, without relying on the code transforming a %d. This code would only transform a single %d as well, and if they had any other formatting symbols before/after, they would also end up with a bad filename.

I will not block this PR, however I don't personally approve of this "magic".
(To clarify, I do like this feature. Overriding the filename would be great. I just don't particularly care for this part.)

Comment (jolheiser) [reply cmd/dump.go]: Sorry for the "dupe" comment notification, I forget that comments in the file section don't show up in conversation unless it's marked as a review.

Comment (glaszig) [reply cmd/dump.go]: i see your point and generally agree. but if you don't submit a custom file name with the `--file` argument, it'll use the default and the default contains `%d`. so, this check is necessary.
we could change the default to not include `%d` but that may break existing setups?

Comment (jolheiser) [reply cmd/dump.go]: Or change the default value to 
```go
Value: fmt.Sprintf("gitea-dump-%d.zip", time.Now().Unix())
```

Comment (techknowlogick) [reply cmd/dump.go]: some pseudo code could be:

```
filename := fmt.Sprintf("gitea-dump-%d.zip", time.Now().Unix())
if ctx.String("file") {
    filename = ctx.String("file")
}
```
and don't do fmt.Sprintf on what is passed in via CLI

Comment (glaszig) [reply cmd/dump.go]: > `Value: fmt.Sprintf("gitea-dump-%d.zip", time.Now().Unix())`

damn. that's brilliant. i'll do that. thank you, sir!

Comment (jolheiser) [reply cmd/dump.go]: Either that or what @techknowlogick said. Whichever seems cleaner in the code. :+1:

Comment (glaszig) [reply cmd/dump.go]: implemented @jolheiser's approach.

**Review by jolheiser — APPROVED** [review]

File: cmd/dump.go (hunk 3 above)
Comment (jolheiser) [reply cmd/dump.go]: :tada: 
Thanks again for the PR!

**Review by techknowlogick — APPROVED** [review]

**Pull request merged**